src/client.rs

```diff
@@ -1099,10 +1103,16 @@ impl<T: Read + Write> Session<T> {
 }
 
 impl<T: Read + Write> Connection<T> {
-    fn read_greeting(&mut self) -> Result<()> {
+    /// Read the greeting from the connection. Needs to be done after `connect`ing.
+    pub fn read_greeting(&mut self) -> Result<Vec<u8>> {
+        if self.greeting_read {
```

Comment:
I think I would just make this `assert!(!self.greeting_read)`. This shouldn't generally be something that only happens _some_ of the time.

Reply:
fixed